tensorboard/components/experimental/plugin_util/plugin-host.ts

```diff
@@ -0,0 +1,100 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {IPC, MessageType, PayloadType, MessageCallback} from './message.js';
+
+const portIPCs = new Set<IPC>();
+const VERSION = 'experimental';
+const listeners = new Map<MessageType, MessageCallback>();
+
+// TODO(@psybuzz): replace this and the port cleanup logic in broadcast() with
+// a MutationObserver to notify us when iframes disconnect.
+const ipcToFrame = new Map<IPC, HTMLIFrameElement>();
+
+// The initial Window-level listener is needed to bootstrap only.
+// All further communication is done over MessagePorts.
+window.addEventListener('message', (event) => {
+  if (event.data !== `${VERSION}.bootstrap`) return;
+  const port = event.ports[0];
+  if (!port) return;
+  const frame = event.source ? event.source.frameElement : null;
+  if (!frame) return;
+  onBootstrap(port, frame as HTMLIFrameElement);
+});
+
+function onBootstrap(port: MessagePort, frame: HTMLIFrameElement) {
+  const portIPC = new IPC(port);
+  portIPCs.add(portIPC);
+  ipcToFrame.set(portIPC, frame);
+  port.start();
+
+  for (const [type, callback] of listeners) {
+    portIPC.listen(type, callback);
+  }
+}
+
+function _broadcast(
+  type: MessageType,
+  payload: PayloadType
+): Promise<PayloadType[]> {
+  for (const ipc of portIPCs) {
+    if (!ipcToFrame.get(ipc).isConnected) {
+      portIPCs.delete(ipc);
+      ipcToFrame.delete(ipc);
+    }
+  }
+
+  const promises = [...portIPCs].map((ipc) => ipc.sendMessage(type, payload));
+  return Promise.all(promises);
+}
+
+function _listen(type: MessageType, callback: MessageCallback) {
+  listeners.set(type, callback);
+  for (const ipc of portIPCs) {
+    ipc.listen(type, callback);
+  }
+}
+
+function _unlisten(type: MessageType) {
+  listeners.delete(type);
+  for (const ipc of portIPCs) {
+    ipc.unlisten(type);
+  }
+}
+
+export const broadcast = _broadcast;
```

Comment:
Is the purpose of this pattern (as opposed to just `export function broadcast(...) {...}`) to prevent monkey-patching?  Just curious if I should adopt this generally.

Reply:
The underscored version is just there to allow us to export it via es6 (which is used/imported by the test) and export it under a namespace, at the same time.